### PR TITLE
kinetis: Add LPM implementation

### PR DESCRIPTION
**Pull request by gebart**
~~This PR depends on the atomic counters in #2321~~ - merged

Add support for Kinetis power modes: WAIT, STOP, VLPS, LLS.

Also included is a refactoring of the UART driver to allow low power modes while still being responsive to RX bytes and prevent deep sleep modes when the last TX byte has not yet been transmitted on the line. This requires using TX interrupts to detect when a transmission is done.

The deepest possible hardware sleep mode will be dynamically selected when lpm_set(LPM_IDLE).

Missing power modes: 
- VLLSx (requires substantial work in the startup code)
- VLPW, VLPR (needs some updates in the bus drivers to reconfigure bus speeds for low power mode)

Tested on:
- k60/mulle

### CODE REVIEW

**Comment by gebart** [comment]
Added a small bug fix for setting up the NVIC when initializing blocking mode.

**Comment by gebart** [comment]
Rebased since #2321 was merged.

**Comment by gebart** [comment]
ping @jfischer-phytec-iot 

Ready for review

**Comment by jfischer-phytec-iot** [comment]
I will adapt and test it on pba-d-01-kw2x.

**Comment by gebart** [comment]
I should add a DSB instruction before the WFI.

**Comment by gebart** [comment]
rebased, added DSB before WFI

**Comment by jfischer-phytec-iot** [comment]
@gebart a0699b0 makes UART broken on the kw2x, the interrupts (from the uart) will be no longer triggered. If I comment out https://github.com/gebart/RIOT/commit/a0699b0032835f3c5a99b87b5ee49edc047c85d1#diff-0f87a847fbd98c6b77134bcd0c723ee5R366 , uart interrupts works again. I still have not figured out why.

**Comment by gebart** [comment]
@jfischer-phytec-iot can you try again with the newly rebased version?
I fixed a bug where the system would deadlock if `printf` was called with interrupts disabled.

Did UART TX still work when you commented out the line you mentioned?

**Comment by jfischer-phytec-iot** [comment]
@gebart Well, I will try during the day.

**Comment by jfischer-phytec-iot** [comment]
@gebart I have tested fdaf098 on kw2x and k64f, uart works :fireworks: . I will test LPM tomorrow.

File: cpu/kinetis_common/include/kinetis_lpm.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+/** @brief LLWU wake up edge trigger modes for external pin wake-ups */
+typedef enum kinetis_lpm_edge {
+    /** Disable external pin wake up */
+    KINETIS_LPM_EDGE_DISABLE = 0b00,
+    /** Wake on rising edge */
+    KINETIS_LPM_EDGE_RISING = 0b01,
+    /** Wake on falling edge */
+    KINETIS_LPM_EDGE_FALLING = 0b10,
+    /** Wake on any logic change */
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/include/kinetis_lpm.h]: over-commented, the identifiers are self-explanatory

File: cpu/kinetis_common/lpm_arch.c [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "arch/lpm_arch.h"
+#include "kinetis_lpm.h"
+
+/*
+ * Counting semaphores for inhibiting undesired power modes.
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/lpm_arch.c]: over-commented, you have already commented in kinetis_lpm.h

File: cpu/kinetis_common/lpm_arch.c [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+ */
+atomic_int_t kinetis_lpm_inhibit_stop_sema = {0};
+atomic_int_t kinetis_lpm_inhibit_vlps_sema = {0};
+atomic_int_t kinetis_lpm_inhibit_lls_sema  = {0};
+
+static inline void wait(void)
+{
+    /* Clear the SLEEPDEEP bit to make sure we go into WAIT (sleep) mode instead
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/lpm_arch.c]: over-commented,  /* wait for outstanding writes and go into WAIT (sleep) mode */ is suffiecient
There are 6 lines of comments and 3 lines of plain code :-)

File: cpu/kinetis_common/lpm_arch.c [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+ *
+ * @param[in] stopmode     The chosen STOP mode
+ *
+ * @see Kinetis CPU reference manual, chapter System Mode Controller, Power Mode
+ * Control register (SMC_PMCTRL), field STOPM
+ */
+static inline void stop(uint8_t stopmode)
+{
+    volatile uint8_t dummy;
+    /* Set the SLEEPDEEP bit to enable deep sleep modes (STOP) */
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/lpm_arch.c]: the function is over-commented, /* wait for outstanding writes and go into STOP (deep sleep) mode */ is enough.

File: cpu/kinetis_common/lpm_arch.c [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    /* WFI instruction will start entry into STOP mode */
+    __WFI();
+}
+
+/** @todo Implement Kinetis power modes: VLPR, VLPW, VLLSx */
+
+static void kinetis_low_power_mode(void)
+{
+    if (kinetis_lpm_inhibit_stop_sema.value != 0) {
+        /* STOP inhibited, go to WAIT mode */
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/lpm_arch.c]: s/wait/go_to_wait_mode/ and the comment is not necessary.

File: cpu/kinetis_common/lpm_arch.c [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        wait();
+    }
+    else if (kinetis_lpm_inhibit_vlps_sema.value != 0) {
+        /* VLPS inhibited, go to normal STOP mode */
+        stop(KINETIS_POWER_MODE_NORMAL);
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/lpm_arch.c]: s/stop/go_to_stop_mode/ ....

File: cpu/kinetis_common/lpm_arch.c [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        stop(KINETIS_POWER_MODE_LLS);
+    }
+}
+
+void lpm_arch_init(void)
+{
+    /* Setup Low Leakage Wake-up Unit (LLWU) */
+    LLWU_UNLOCK();           /* Enable LLWU clock gate */
+    KINETIS_PMPROT_UNLOCK(); /* Enable all available power modes */
+    /* Enables wake on LPTMR, to let the hwtimer take the chip out of sleep */
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/lpm_arch.c]: over-commented, the identifier of the function is self-explanatory, the same L104 !

**Comment by jfischer-phytec-iot** [comment]
@gebart 
I tested kinetis lpm with kw2x. Energy consumption was reduced by half (periph_uart_int  and vtimer_msg)

Is it (maybe) not problematic that a driver (e.g. uart) sets which power mode is possible? It is somehow not right.

**Comment by gebart** [comment]
@jfischer-phytec-iot 
>Is it (maybe) not problematic that a driver (e.g. uart) sets which power mode is possible? It is somehow not right.

I agree, but in cpu/periph interface drivers the developer must know what power modes are available and what power modes allow what functionality in the driver. For example, we know that the UART always becomes static in STOP and lower modes, resulting in garbage on the receiving end if we are in the middle of a transmission when STOP mode is entered, therefore we can place an inhibitor on STOP in that driver.

We can't really remove the LPM inhibitors from the drivers either, since that would mean that the LPM driver would have to know the internals of all other drivers instead, in order to figure out which modes are allowed.

**Comment by gebart** [comment]
Merged kw2x implementation from @jfischer-phytec-iot 

@jfischer-phytec-iot Do you have an idea of which of the defines and enums should be moved to `kinetis_common`?

**Comment by jfischer-phytec-iot** [comment]
@gebart sorry for delay, as I see are the UART-RX-Pins also Wakeup Pins. It should be possible (also with rxcallbacks) to go into deeper sleep modes. You have to take care of the MCG module, I'll try it out and report tomorrow.

**Comment by gebart** [comment]
As far as I can see in the K60 reference manual, the UART module is completely static in LLS and lower. Wake on edge detect is only available in VLPS and above. (Chapter 7 Power Management Table 7-2. Module operation in low power modes)

**Comment by jfischer-phytec-iot** [comment]
> Do you have an idea of which of the defines and enums should be moved to kinetis_common?

As far as I can see, the following bit fields are identical:
+ STOP mode bitfield values
+ RUN mode bitfield values
+ LLWU wake up sources (module and pins)
+ VLLS Mode Control

**Comment by jfischer-phytec-iot** [comment]
@gebart I believe that lpm_arch_init was never called :rotating_light: :smile:

**Comment by gebart** [comment]
There's a problem with this PR, the application hangs when I don't have a debugger connected, but works fine with the debugger attached. Something to do with the UART. I am investigating this.

**Comment by jfischer-phytec-iot** [comment]
@gebart 
I can confirm that.

File: cpu/kinetis_common/lpm_arch.c [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+void lpm_arch_end_awake(void)
+{
+    /* TODO */
+}
+
+void kinetis_lpm_enable_wakeup_module(const uint8_t module) {
+    /* Use bit-banding for thread-safety instead of disabling interrupts */
+    BITBAND_REG32(LLWU->ME, module) = 1;
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/lpm_arch.c]: LLWU.x are 8bit register

File: cpu/kinetis_common/lpm_arch.c [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+void lpm_arch_init(void)
+{
+    /* Setup Low Leakage Wake-up Unit (LLWU) */
+    LLWU_UNLOCK();           /* Enable LLWU clock gate */
+    KINETIS_PMPROT_UNLOCK(); /* Enable all available power modes */
+    /* Enables wake on LPTMR, to let the hwtimer take the chip out of sleep */
+    kinetis_lpm_enable_wakeup_module(KINETIS_LPM_WAKEUP_MODULE_LPTMR);
+
+    NVIC_EnableIRQ(KINETIS_LLWU_IRQ); /* Enable LLWU interrupt */
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/lpm_arch.c]: isr_llwu(void) should be redefined below.

File: cpu/kinetis_common/lpm_arch.c [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+        stop(KINETIS_POWER_MODE_NORMAL);
+    }
+    else if (kinetis_lpm_inhibit_lls_sema.value != 0) {
+        /* LLS inhibited, go to VLPS mode */
+        stop(KINETIS_POWER_MODE_VLPS);
+    }
+    else {
+        /* go to LLS mode */
+        stop(KINETIS_POWER_MODE_LLS);
+    }
```
Comment (jfischer-phytec-iot) [inline cpu/kinetis_common/lpm_arch.c]: before exit, MCG mode (e.g. PEE) should be restarted, like kinetis_mcg_set_mode(KINETIS_MCG_PEE);

File: cpu/kinetis_common/lpm_arch.c (hunk 9 above)
Comment (gebart) [inline cpu/kinetis_common/lpm_arch.c]: That's probably it! Thank you, I will test this tomorrow morning or on
Monday.

**Comment by jfischer-phytec-iot** [comment]
@gebart
I have it so far that it wakes up from LLS and VLPs, but the 115200 baud are just too much and leads to errors. See also  https://community.freescale.com/thread/327741


I'll make a patch for the PLL (mcg).

**Comment by jfischer-phytec-iot** [comment]
@gebart my testing branch https://github.com/jfischer-phytec-iot/RIOT/tree/gebart%40pr/kinetis-lpm

**Comment by gebart** [comment]
@jfischer-phytec-iot can you test if 57600 baud works without corruption?

**Comment by jfischer-phytec-iot** [comment]
@gebart Unfortunately no difference to 115200 baud. It's something with UART. Actually, it should be enough time for the Start-Bit detection: "The receiver has an acquisition rate of 16 samples per bit time." (See Baud rate generation).

**Comment by jfischer-phytec-iot** [comment]
wakeup

CH1: LOW - STOP, HIGH - RUN
CH2: UART.RX

**Comment by gebart** [comment]
@jfischer-phytec-iot did you disable interrupts before you call WFI when you got that scope reading?

The CPU will still be awoken from sleep if interrupts are disabled but it will not jump to the ISR until interrupts are reenabled again.

**Comment by jfischer-phytec-iot** [comment]
@gebart The interrupts are enabled. The measure appears to be correct. The CPU runs at 4MHz just after wakeup (PBE mode). 50µs are about 8 CPU cycles and before the PLL output is selected again goes a little time, then it's too late. Well, I want to try in another MCG mode. On your board it should behave differently.

:hammer:  I wish we had a sleep command for the shell.

**Comment by gebart** [comment]
@jfischer-phytec-iot 
> I wish we had a sleep command for the shell.
Good idea, I think we should add it. However, should that command power down all peripheral devices before putting the system to sleep? How do we wake the system again? hwtimer?

**Comment by gebart** [comment]
I'm having some trouble with the LPM implementation. On Contiki I can get the entire board down to around 40 µA with all on-board components powered and the CPU in LLS, and around 200 µA in VLPS, but currently I'm seeing 1 mA with RIOT, so there is some part which is not being properly powered down yet.. I'm suspecting the radio, but it seem to be a greater difference than what the transceiver would account for in TRX_OFF state (data sheet lists the current consumption at 450 µA, typically, in TRX_OFF).

**Comment by jfischer-phytec-iot** [comment]
>Good idea, I think we should add it. However, should that command power down all peripheral devices before putting the system to sleep? How do we wake the system again? hwtimer?

I think the activity at the UART RX line can serve as a wakeup signal. It will lose a character ("wakeup character"), but I think it's ok as long as UART will be used for the shell only. And in case "RIOT as borderrouter", LPM can be switched off completely.

**Comment by jfischer-phytec-iot** [comment]
@gebart I've implemented a lpm command (lpm [sleep, powerdown, off]) for the shell and now I see that without proper power manager, we will not get anywhere.

**Comment by jfischer-phytec-iot** [comment]
@gebart 
https://github.com/jfischer-phytec-iot/RIOT/commits/gebart%40pr/kinetis-lpm
https://github.com/jfischer-phytec-iot/RIOT/commit/3034040ca40c0e4662e2ff88423b555fe0d59e59
look at it when you have time.

**Comment by gebart** [comment]
I managed to get everything down to 100 µA in VLPS, and around 50 µA in LLS, using the hello-world example (so no processes are running except for the LPTMR overflow ISR handler every 2 seconds). To get low power I had to add initialization code for some of the on-board devices (flash memory and radio transceiver) to trigger low power modes during boot. It does not seem to affect the device initialization performed by the at86rf231 driver.

@jfischer-phytec-iot Do you have any opinions on this method? Hard coding small init sequences for on board devices inside the `board_init` function, even when the actual driver for the device has not been enabled in the build.

@jfischer-phytec-iot I looked at your suggestion, it was something like that I had in mind when I opened #2927. I like it, do you think we should merge this first and then do a separate PR for the power manager, or do the power manager first, or combine the two in one PR? I think this one (#2605) is pretty close to being mergable, but we will still have to rewrite it when we introduce the power manager.

**Comment by zhuoshuguo** [comment]
@gebart Ping. A kind reminder for rebase.

**Comment by gebart** [comment]
on hold until xtimer catches up

**Comment by miri64** [comment]
With #6160 merged: Is this still relevant?

**Comment by smlng** [comment]
`lpm` was removed in #6160, as @miri64 already said. Hence, +1 for closing this one

**Comment by PeterKietzmann** [comment]
Guess it's outdated. Will close with memo label set. @gebart please reopen if I'm mistaken.

**Comment by maribu** [comment]
This PR is still referred to from the `mulle` boards doc. Can someone familiar with the board update the doc?

**Pull request closed without merging**